src/debug_utils-inl.h

```diff
@@ -0,0 +1,91 @@
+#ifndef SRC_DEBUG_UTILS_INL_H_
+#define SRC_DEBUG_UTILS_INL_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "debug_utils.h"
+
+#include <type_traits>
+
+namespace node {
+
+struct ToStringHelper {
+  template <typename T>
+  static std::string Convert(
+      const T& value,
+      std::string(T::* to_string)() const = &T::ToString) {
+    return (value.*to_string)();
+  }
+  template <typename T,
+            typename test_for_number = typename std::
+                enable_if<std::is_arithmetic<T>::value, bool>::type,
+            typename dummy = bool>
+  static std::string Convert(const T& value) { return std::to_string(value); }
+  static std::string Convert(const char* value) { return value; }
+  static std::string Convert(const std::string& value) { return value; }
+  static std::string Convert(bool value) { return value ? "true" : "false"; }
+};
+
+template <typename T>
+std::string ToString(const T& value) {
+  return ToStringHelper::Convert(value);
+}
+
+inline std::string SPrintFImpl(const char* format) {
+  return format;
+}
+
+template <typename Arg, typename... Args>
+std::string COLD_NOINLINE SPrintFImpl(  // NOLINT(runtime/string)
+    const char* format, Arg&& arg, Args&&... args) {
+  const char* p = strchr(format, '%');
+  if (p == nullptr) return format;
```

Comment:
Ideally, this would e.g. abort when there are still arguments left. (Bonus points if you can make that work at compile time!)

And vice versa, the non-template version could/should error when there are still format specifiers except `%%` left.

Reply:
> Ideally, this would e.g. abort when there are still arguments left.

> And vice versa, the non-template version could/should error when there are still format specifiers except `%%` left.

I’ve had that in here for an earlier version of the code (for both case) but realized it would be somewhat unhelpful because we mostly use this for debugging at this point. I’ve added it back now but I still feel a bit unsure whether this is the most helpful thing to do here.

> (Bonus points if you can make that work at compile time!)

I’d love that but I wouldn’t quite know how to do that without making usage less ergonomic…